torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+  CacheNode& operator=(const CacheNode&) = delete;
+  CacheNode& operator=(CacheNode&&) = delete;
+
+  bool check_dynamic_sizes(AutogradCompilerCall& call) {
+    /*
+    We start off by assuming everything is static, then we mark things
+    as dynamic when we see them change.  This function:
+      1) Checks for a cache hit
+      2) Updates expected_sizes to track what is dynamic
+      3) Populates call.dyn_size_inputs by filtering call.all_size_inputs
+    */
+    bool cache_hit = compiled_fn.get() != nullptr;
+    auto len = call.all_size_inputs.size();
+    const SizeInput* data = call.all_size_inputs.data();
+    if (expected_sizes.empty()) {
+      expected_sizes.reserve(len);
+      for (const auto i : c10::irange(len)) {
+        expected_sizes.emplace_back(data[i]);
+      }
+    }
+
+    TORCH_CHECK(expected_sizes.size() == call.all_size_inputs.size());
+    for (const auto i : c10::irange(len)) {
+      auto& expected = expected_sizes[i];
+      if (expected.dyn_type == SizeInput::DYNAMIC ||
+          expected.value != data[i].value) {
+        cache_hit = cache_hit && expected.dyn_type == SizeInput::DYNAMIC;
+        if (expected.value != data[i].value) {
+          expected = SizeInput(SizeInput::DYNAMIC, data[i].value);
+        }
+        if (call.dyn_size_inputs.empty()) {
+          call.dyn_size_inputs.reserve(len);
+        }
+        call.dyn_size_inputs.emplace_back(data[i].value);
+      }
+    }
+
+    if (!cache_hit) {
+      compiled_fn = nullptr;
+    }
+    return cache_hit;
+  }
+
+  PyObject* wrap_dynamic_inputs() {
+    size_t dynamic_count = 0;
+    size_t idx = 0;
+    for (const auto& i : expected_sizes) {
+      if (i.dyn_type == SizeInput::DYNAMIC) {
+        ++dynamic_count;
+      }
+    }
+    PyObject* pyinput = PyTuple_New(dynamic_count);
+    for (const auto& i : expected_sizes) {
+      if (i.dyn_type == SizeInput::DYNAMIC) {
+        PyTuple_SET_ITEM(pyinput, idx++, PyLong_FromSsize_t(i.value));
+      }
+    }
+    TORCH_CHECK(idx == dynamic_count);
+    return pyinput;
+  }
+
+  std::vector<c10::optional<SymInt>> unwrap_dynamic_inputs(PyObject* pyresult) {
+    TORCH_CHECK(PyList_CheckExact(pyresult));
+    size_t idx = 0;
+    size_t result_len = PyList_GET_SIZE(pyresult);
+    std::vector<c10::optional<SymInt>> result;
+    result.reserve(expected_sizes.size());
+    for (const auto& i : expected_sizes) {
+      if (i.dyn_type == SizeInput::DYNAMIC) {
+        TORCH_CHECK(idx < result_len);
+        result.emplace_back(
+            py::cast<c10::SymInt>(PyList_GET_ITEM(pyresult, idx++)));
+      } else {
+        result.emplace_back();
+      }
+    }
+    TORCH_CHECK(idx == result_len && result.size() == expected_sizes.size());
+    return result;
+  }
+
+  // TODO(jansel): benchmark map vs unordered_map
+  std::unordered_map<CacheKey, std::unique_ptr<CacheNode>> next;
+  std::vector<CacheKeyBuffer> key_storage;
+  std::vector<SizeInput> expected_sizes;
+  THPObjectPtr compiled_fn;
```

Comment:
It seems like this would be better as a SafePyObject? Then you don't have to manually disarm the deallocation at shutdown.

Reply:
SafePyObject does not support assignment.

Reply:
Oh yeah, the easy workaround for this problem is to put it in a shared_ptr LOL

Reply:
Yeah, at that point it would be more code (and an extra malloc) than just disarm on shutdown.  We also don't need the mutli-interpreter stuff since this lives in the python `.so`.